cmake/extensions.cmake

```diff
@@ -2282,6 +2282,29 @@ Relative paths are only allowed with `-D${ARGV1}=<path>`")
   endif()
 endfunction()
 
+# Usage:
+#   zephyr_file_rename(<name> <newname>)
+#
+# Zephyr file function extension.
+# This function renames the file <name> to <newname> if content of '<name>'
+# differs from content of <newname>.
+#
+# If file should be renamed unconditionally, please use CMake function
+# 'file(RENAME ...)' instead.
+#
+# The file '<name>' are always removed after this function returns.
+function(zephyr_file_rename name newname)
+  execute_process(
+    COMMAND ${CMAKE_COMMAND} -E compare_files ${name} ${newname}
+    RESULT_VARIABLE compare_result
+  )
+  if(NOT ${compare_result} EQUAL 0)
+    file(RENAME ${name} ${newname})
+  else()
+    file(REMOVE ${name})
+  endif()
```

Comment:
```
if(CMAKE_VERSION GREATER_OR_EQUAL 2.21.0)
file(COPY_FILE ${name} ${newname} ONLY_IF_DIFFERENT)
endif()

```

I'm concerned about the runtime overhead of "cmake -E compare_files" being invoked half a dozen times. I suspect
file(COPY_FILE will be faster (it probably doesn't create a new process, which is costly (especially on Win) ).

I suggest we use COPY_FILE when available.

COPY_FILE doesn't always delete the old file, which is
unfortunate, but acceptable if it makes the build go faster.
I guess we could add a "rm ${name}" to reduce the size of the build dir.

Reply:
done, when using `>=3.21` we use built-in function.